### ISSUE TITLE
Improve Playground Samples for Language Providers

### ISSUE THREAD

**Issue by hawkerm** [microsoft/monaco-editor#1890]
Language providers are very powerful, but not all the ones supported in Monaco have examples in the Playground to show how to use them and how they work.  Some of the existing samples provide a blank document, so it's unclear on what to do in order to see the key results.

It'd be great for more comments/information for each of the provider samples (like the symbol provider) be given as well as new samples for thinks like code actions, declaration, definition, documentHighlight, link, rename, onTypeFormattingEdit, etc...

**Comment by rcjsuen** [comment]
> It'd be great for more comments/information for each of the provider samples (like the symbol provider)

I would say it might be better to just rewrite the symbol provider example given that it's not reflective of what is in the document at all. Perhaps it could be "improved" to clearly state that it is for a visual check of what the icons look like rather than an accurate representation of a correctly implemented symbol provider. 🤔

**Comment by rcjsuen** [comment]
@hawkerm Are you confused about any specific language feature provider?

**Comment by hawkerm** [comment]
@rcjsuen it's partly about understanding the features themselves (if they linked to any VS Code documentation on the feature for instance that could be helpful). I think a couple of places in the Monaco API docs do this, but not all.

The playground is a nice way to explore the APIs, test them, and understand the mapping of all the parameters to their function in modifying behaviors of the editor. It makes it really easy to tweak values and see different outcomes.

I've been writing a [C# UWP wrapper](https://github.com/hawkerm/monaco-editor-uwp/), now used in a few apps, but it's difficult without a good sample of some features to translate them over and have a known test case to compare against.

**Comment by rcjsuen** [comment]
> new samples for thinks like code actions, declaration, definition, documentHighlight, link, rename, onTypeFormattingEdit, etc...

@hawkerm Do you have a ranking of which providers you need samples for or do you not care and just want every provider to have a sample?

**Comment by acao** [comment]
while the monaco editor playground shows a good provider example or two, there could be more. it would be especially nice to have an example that shows how to introduce a custom language service worker, that leverages the worker in the providers. in addition to one with amd, it would be nice to have an example of this using `esm` modules, as i can't seem to figure out how to specify `moduleId` for `createWebWorker`

**Comment by rcjsuen** [comment]
> it would be especially nice to have an example that shows how to introduce a custom language service worker, that leverages the worker in the providers.

@acao What are the advantages of using a language service worker?

I don't use any workers in my [Dockerfile editor](https://rcjsuen.github.io/dockerfile-language-service/) but maybe that's because I don't have a complicated use case?

**Comment by hawkerm** [comment]
@rcjsuen I already have Color, CompletionItem, and Hover figured out, so I'd probably order the rest as:

1. CodeAction
2. DocumentFormattingEdit / DocumentRangeFormattingEdit / OnTypeFormattingEdit
3. CodeLens
4. Definition / Declaration / Implementation / Reference / TypeDefinition
5. Rename
6. DocumentHighlight / DocumentSymbol
7. SelectionRange (does this help define boundaries to snap selections to?)
8. DocumentSemanticTokens / DocumentRangeSemanticTokens (Not really sure what this is at all)
9 .SignatureHelp (not sure if this is useful for XML?)
10. LinkProvider (there's basic support for hyperlinks already)
11. FoldingRange (the XML folding works fine for my case already)

**Comment by rcjsuen** [comment]
@hawkerm [This document](https://code.visualstudio.com/api/language-extensions/programmatic-language-features) may help if you haven't already seen it.

**Comment by buremba** [comment]
[monaco-json](https://github.com/microsoft/monaco-json) may be a good example of a language implementation for Monaco. It lacks some of the features such as code actions and go-to-definition though.

**Comment by redexp** [comment]
@rcjsuen 

> [This document](https://code.visualstudio.com/api/language-extensions/programmatic-language-features) may help if you haven't already seen it.

@rcjsuen This link should be in top menu of [Monaco Editor site](https://microsoft.github.io/monaco-editor/index.html)

**Comment by hawkerm** [comment]
Was trying to get a CodeAction sample up and running, but not sure what I'm missing. Not enough details in the VS Code specs as they implement it from a language provider stand-point.

Here's what I have so far in the playground:

```js
var editor = monaco.editor.create(document.getElementById("container"), {
    value: "{\n\t\"dependencies\": {\n\t\t\"hello\": \"world\"\n\t}\n}\n",
    language: "json"
});

var commandId = editor.addCommand(0, function () {
    // services available in `ctx`
    alert('my command is executing!');
}, '');

monaco.languages.registerCodeActionProvider('json', {
    provideCodeActions: function (model, range, context, token) {
        // Should have an if to see if we're in the right spot?
        return {
            actions: [
                {
                    title: "This is a test action.",
                    kind: "quickfix",
                    command: {
                        id: commandId,
                        title: "A Command Here"
                    },
                    diagnostics: [
                        {
                            code: "2344",
                            message: "A warning",
                            severity: "Warning",
                            startLineNumber: 3,
                            startColumn: 1,
                            endLineNumber: 3,
                            endColumn: 12,
                        }
                    ],
                    edit: {
                        edits: [
                            {
                                edit: {
                                    text: "test text",
                                    range: {
                                        startLineNumber: 3,
                                        startColumn: 1,
                                        endLineNumber: 3,
                                        endColumn: 12
                                    }
                                }
                            }
                        ]
                    }
                }
            ],
            dispose: () => {}
        };
    },
});
```

This shows an action:
![image](https://user-images.githubusercontent.com/8959496/127044233-d746fd5e-fbfa-46ba-94f8-c81174086938.png)

But then doesn't do anything, looks like I don't know how to make an `edit` either:
![image](https://user-images.githubusercontent.com/8959496/127044294-642ee4ca-f08d-4bad-ade7-92ad139e8030.png)

**Comment by rcjsuen** [comment]
> @rcjsuen
> 
> > [This document](https://code.visualstudio.com/api/language-extensions/programmatic-language-features) may help if you haven't already seen it.
> 
> This link should be in top menu of [Monaco Editor site](https://microsoft.github.io/monaco-editor/index.html)

@redexp Yeah, it wouldn't hurt. I don't work for Microsoft though so I have no say here. You could try opening a pull request with your suggestion by modifying the [`index.html` file](https://github.com/microsoft/monaco-editor/blob/46bc7435b7509a455252b8678102e2c32130dc03/website/index.html) and see what feedback you get.

> Was trying to get a CodeAction sample up and running, but not sure what I'm missing.

@hawkerm Your `edit` does not have a `resource` attribute specified so Monaco does not know which model to change.

- [`CodeAction`](https://microsoft.github.io/monaco-editor/api/interfaces/monaco.languages.codeaction.html)
- [`WorkspaceEdit`](https://microsoft.github.io/monaco-editor/api/interfaces/monaco.languages.workspaceedit.html)
- [`WorkspaceTextEdit`](https://microsoft.github.io/monaco-editor/api/interfaces/monaco.languages.workspacetextedit.html)
- [`resource`](https://microsoft.github.io/monaco-editor/api/interfaces/monaco.languages.workspacetextedit.html#resource)

```JavaScript
var editor = monaco.editor.create(document.getElementById("container"), {
    value: "{\n\t\"dependencies\": {\n\t\t\"hello\": \"world\"\n\t}\n}\n",
    language: "json"
});

var commandId = editor.addCommand(0, function () {
    // services available in `ctx`
    alert('my command is executing!');
}, '');

monaco.languages.registerCodeActionProvider('json', {
    provideCodeActions: function (model, range, context, token) {
        // Should have an if to see if we're in the right spot?
        return {
            actions: [
                {
                    title: "This is a test action.",
                    kind: "quickfix",
                    command: {
                        id: commandId,
                        title: "A Command Here"
                    },
                    diagnostics: [
                        {
                            code: "2344",
                            message: "A warning",
                            severity: "Warning",
                            startLineNumber: 3,
                            startColumn: 1,
                            endLineNumber: 3,
                            endColumn: 12,
                        }
                    ],
                    edit: {
                        edits: [
                            {
                                edit: {
                                    text: "test text",
                                    range: {
                                        startLineNumber: 3,
                                        startColumn: 1,
                                        endLineNumber: 3,
                                        endColumn: 12
                                    }
                                },
                                // this was missing which made your edit
                                // unrecognizable, exception thrown
                                resource: editor.getModel().uri
                            }
                        ]
                    }
                }
            ],
            dispose: () => {}
        };
    },
});
```

**Comment by hawkerm** [comment]
Thanks @rcjsuen I thought I had still had that in there, seems to be working in Monaco Playground, though I'm on 21.3 still and having issues, so opening another bug.
